pumi/pumi_mesh.cc

```diff
@@ -0,0 +1,624 @@
+      for (int i=0; i < na; ++i)
+      {
+        if ( ! m->hasTag(adjacent[i],tag))
+        {
+          m->setIntTag(adjacent[i],tag,&dummy);
+          affected[dimension].push_back(adjacent[i]);
+        }
+        Copies remotes;
+        m->getRemotes(adjacent[i],remotes);
+        APF_ITERATE(Copies,remotes,rit)
+          PCU_COMM_PACK(rit->first,rit->second);
+        if (m->hasMatching())
+        {
+          apf::Matches matches;
+          m->getMatches(adjacent[i],matches);
+          for (size_t j=0; j < matches.getSize(); ++j)
+            PCU_COMM_PACK(matches[j].peer,matches[j].entity);
+        }
+      }//downward adjacent loop
+    }//upward affected loop
+    PCU_Comm_Send();
+    while (PCU_Comm_Receive())
+    {
+      pMeshEnt entity;
+      PCU_COMM_UNPACK(entity);
+      if ( !m->hasTag(entity,tag))
+      {
+        m->setIntTag(entity,tag,&dummy);
+        affected[dimension].push_back(entity);
+      }
+    }
+    APF_ITERATE(EntityVector,affected[dimension],it)
+      m->removeTag(*it,tag);
+  }//dimension loop
+  m->destroyTag(tag);
+}
+
+static Parts distr_makeResidence(Parts& parts)
+{
+  Parts r;
+  APF_ITERATE(Parts, parts, pit)
+    r.insert(*pit);
+  return r;
+}
+
+// *********************************************************
+static void distr_updateResidences(pMesh m,
+    Distribution* plan, EntityVector affected[4])
+// *********************************************************
+{
+  int maxDimension = m->getDimension();
+  pMeshEnt e;
+
+  apf::MeshIterator* it = m->begin(maxDimension);
+  int i=0;
+  while ((e = m->iterate(it)))
+  {
+    if (plan->parts_vec[i].size()>0) 
+    {
+      Parts res = distr_makeResidence(plan->parts_vec[i]);
+      m->setResidence(e,res);
+    }
+    ++i;
+  }
+  m->end(it);
+
+  for (int dimension = maxDimension-1; dimension >= 0; --dimension)
+  {
+    PCU_Comm_Begin();
+    APF_ITERATE(EntityVector,affected[dimension],it)
+    {
+      pMeshEnt entity = *it;
+      Parts newResidence;
+      Up upward;
+      m->getUp(entity, upward);
+      for (int ui=0; ui < upward.n; ++ui)
+      {
+        pMeshEnt up = upward.e[ui];
+        Parts upResidence;
+        m->getResidence(up,upResidence);
+        apf::unite(newResidence,upResidence);
+      }
+      m->setResidence(entity,newResidence);
+      Copies remotes;
+      m->getRemotes(entity,remotes);
+      APF_ITERATE(Copies,remotes,rit)
+      {
+        PCU_COMM_PACK(rit->first,rit->second);
+        apf::packParts(rit->first,newResidence);
+      }
+    }
+    PCU_Comm_Send();
+    while(PCU_Comm_Receive())
+    {
+      pMeshEnt entity;
+      PCU_COMM_UNPACK(entity);
+      Parts current;
+      m->getResidence(entity,current);
+      Parts incoming;
+      apf::unpackParts(incoming);
+      apf::unite(current,incoming);
+      m->setResidence(entity,current);
+    }
+  }
+}
+
+// *********************************************************
+void distribute(pMesh m, Distribution* plan)
```

Comment:
Distribution is not the same as "multi-layer" ghosting. It allows to support the multiple element copies and each recognizes off-part copies as remote copies.